web/toast_manager.js

```diff
@@ -0,0 +1,77 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class ToastManager {
+  #toastElement;
+
+  #duration;
+
+  #timeoutID = null;
+
+  #controller = null;
+
+  #boundHide = this.#hide.bind(this);
+
+  constructor(elem, duration) {
+    this.#toastElement = elem;
+    this.#duration = duration;
+  }
+
+  show(action, type) {
+    if (this.#timeoutID) {
+      this.#finalizeTimeout();
+    }
+    if (this.#controller) {
+      this.#finalizeController();
+    }
+    this.#toastElement.setAttribute("data-l10n-args", JSON.stringify({ type }));
+    this.#toastElement.removeAttribute("hidden");
+    this.#timeoutID = setTimeout(this.#boundHide, this.#duration);
+    this.#controller = new AbortController();
+    this.#toastElement
+      .querySelector("#annotationRemovedUndoButton")
+      .addEventListener(
+        "click",
+        () => {
+          action();
+          this.#hide();
+        },
+        { signal: this.#controller.signal }
+      );
+    this.#toastElement
+      .querySelector("#annotationRemovedCloseButton")
+      .addEventListener("click", this.#boundHide, {
+        signal: this.#controller.signal,
+      });
+  }
+
+  #finalizeTimeout() {
+    clearTimeout(this.#timeoutID);
+    this.#timeoutID = null;
+  }
+
+  #finalizeController() {
+    this.#controller.abort();
+    this.#controller = null;
+  }
+
+  #hide() {
+    this.#toastElement.setAttribute("hidden", "");
+    this.#finalizeTimeout();
+    this.#finalizeController();
+  }
```

Comment:
Why do we need separate methods, since that seems unnecessary?
```suggestion
  #hide() {
    this.#container.hidden = true;
    
    if (this.#timeoutId) {
      clearTimeout(this.#timeoutId);
      this.#timeoutId = null;
    }
    this.#controller?.abort();
    this.#controller = null;
  }
```

Reply:
They're necessary because I'm also conditionally calling them at the start of `show` to cleanup any previous timeouts or controllers if they exist.

Reply:
(requested in https://github.com/mozilla/pdf.js/pull/18900#discussion_r1799764474)

Reply:
> They're necessary because I'm also conditionally calling them at the start of `show` to cleanup any previous timeouts or controllers if they exist.

Please just implement this, see also https://github.com/mozilla/pdf.js/pull/18900#discussion_r1800676358 below.

Reply:
Fixed, thanks I'd missed the other comment